### PR TITLE
MAINT: Translate lambertw_scalar into C++

### PR DESCRIPTION
**Pull request by steppi**
#### Reference issue

This PR closes no issues, but this was discussed in the RFC issue gh-19404.

#### What does this implement/fix?

This PR rewrites the scalar kernel of the lambertw function in C++. The translation is as literal as possible, the only C++ism involved is the use of `std::complex`. Since the ufunc machinery in the numpy C API works with C doubles, I had to write a wrapper for C++ implementation which takes the complex argument as an `npy_cdouble`, converts to `std::complex<double>`, and then converts the output back into an `npy_cdouble`. This introduces some overhead, but the impact appears negligible.

`_lambertw.pxd` depended on `_evalpoly.pxd` which contains a polynomial evaluation implementation for polynomials with real coefficients and a complex argument. I've created a C++ implementation in `_evalpoly.h` which is again a direct translation of the corresponding Cython code. `_evalpoly.pxd` remains in place because it is a dependency of `_loggamma.pxd`. I could have made an `extern "C"` wrapper for the C++ code to be used in Cython, but we would lose the ability to inline this call. I think we can just have `_evalpoly.pxd` remain in place until `_loggamma.pxd` is translated.

 All functions are contained entirely in headers in order to facilitate the possible inclusion in other projects, as suggested by @izaid. I've verified that results are identical within machine precision on all test cases in `test_lambertw.py` (they could be completely identical, but I didn't check that).

#### Additional information

@izaid, please take a look and try the implementation out with CUDA. I'm curious how that works. I plan to work through the CUDA C++ programming guide and hope to be able to do stuff like that myself soon too. There's an open issue on CuPy for lambertw https://github.com/cupy/cupy/issues/4450. It would be nice if we could make a PR there as well in the near future.

### CODE REVIEW

**Comment by ilayn** [comment]
> This introduces some overhead, but the impact appears negligible.

I think use of `Union` (even though they are frowned upon by C++ folk) is a good alternative as the size of the types are identical. This cryptic explanation that flies way above my head also says familiar sounding things https://en.cppreference.com/w/cpp/language/reinterpret_cast

We have some union code also in Cython that I don't remember where we put it.

File: scipy/special/_evalpoly.h [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+ *
+ * then coeffs[0] = u_n, coeffs[1] = u_{n - 1}, ..., coeffs[n] = u_0.
+ *
+ * References
+ * ----------
+ * [1] Knuth, "The Art of Computer Programming, Volume II"
+ */
+
+
+#ifndef EVALPOLY_H
```
Comment (izaid) [inline scipy/special/_evalpoly.h:16]: Rather than include guards, we can use `#pragma once` here. See [Pragma once](https://en.wikipedia.org/wiki/Pragma_once). That's more modern, but if someone knows of some weird case it doesn't work please say! I've never experienced any problems with it...

File: scipy/special/_lambertw.h [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+ *     http://code.google.com/p/mpmath/source/browse/trunk/mpmath/functions.py?spec=svn994&r=992
+ * [3] mpmath source code, Subversion revision 994
+ *     http://code.google.com/p/mpmath/source/browse/trunk/mpmath/function_docs.py?spec=svn994&r=994
+
+ * TODO: use a series expansion when extremely close to the branch point
+ * at `-1/e` and make sure that the proper branch is chosen there.
+ */
+
+
+#ifndef LAMBERTW_H
```
Comment (izaid) [inline scipy/special/_lambertw.h:23]: Same point about `#pragma once`.

File: scipy/special/_lambertw.h [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+#define LAMBERTW_H
+
+#include <cmath>
+#include <complex>
+#include <numpy/npy_math.h>
```
Comment (izaid) [inline scipy/special/_lambertw.h:28]: The `#include <numpy/npy_math.h>` is problematic for other array libraries. It may not exist and it certainly won't work with NVCC's JIT. What I'd suggest is we create a new single header file, just for SciPy, that brings together all the special function C++ implementations and any NumPy stuff they need.

**Comment by steppi** [comment]
> > This introduces some overhead, but the impact appears negligible.
> 
> I think use of `Union` (even though they are frowned upon by C++ folk) is a good alternative as the size of the types are identical. This cryptic explanation that flies way above my head also says familiar sounding things https://en.cppreference.com/w/cpp/language/reinterpret_cast
> 
> We have some union code also in Cython that I don't remember where we put it.

Thanks @ilayn. I'd 

> > This introduces some overhead, but the impact appears negligible.
> 
> I think use of `Union` (even though they are frowned upon by C++ folk) is a good alternative as the size of the types are identical. This cryptic explanation that flies way above my head also says familiar sounding things https://en.cppreference.com/w/cpp/language/reinterpret_cast
> 
> We have some union code also in Cython that I don't remember where we put it.

Thanks @ilayn, I'm not averse to using unions here. Strictly speaking it's undefined behavior to use unions this way in C++. Is it supported in all of the compilers we support? I'll check the performance impact locally tomorrow.

File: scipy/special/_lambertw.h [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+
+#include "_evalpoly.h"
+#include "sf_error.h"
```
Comment (izaid) [inline scipy/special/_lambertw.h:31]: The `#include "sf_error.h"` has similar issues as to the NumPy header. Additionally, it's not even a header, as there is a `sf_error.c`. Even if it was a header, it can't work in other array libraries as is, as I think I saw it making calls to the Python C API.

I'm not saying we need to fix this now, probably it should be done in a follow-up PR, but there needs to be thought into how error handling is going to work in general.

File: scipy/special/_lambertw.h (hunk 3 above)
Comment (steppi) [reply scipy/special/_lambertw.h:28]: Thanks @izaid, definitely an oversight on my part. That's exactly how we should do it.

File: scipy/special/_lambertw.h [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+
+inline std::complex<double> lambertw(std::complex<double> z, long k, double tol) {
+    int i;
+    double absz;
+    std::complex<double> w;
+    std::complex<double> ew, wew, wewz, wn;
+
+    if (std::isnan(z.real()) || std::isnan(z.imag())) {
+	return z;
+    } else if (z.real() == INFINITY) {
```
Comment (izaid) [inline scipy/special/_lambertw.h:49]: For constants like `INFINITY`, might be better to use `std::numeric_limits<double>::infinity()`. See [numeric limits](https://en.cppreference.com/w/cpp/types/numeric_limits).

File: scipy/special/_lambertw.h (hunk 4 above)
Comment (steppi) [reply scipy/special/_lambertw.h:31]: Ah, right. The best thing I can think of is to put checks in the wrapper that has the numpy stuff.

File: scipy/special/_lambertw.h [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+	return z + std::complex<double>(0, 2.0*M_PI*static_cast<double>(k));
+    } else if (z.real() == -INFINITY) {
+	return -z + std::complex<double>(0, 2.0*M_PI*static_cast<double>(k) + M_PI);
```
Comment (izaid) [inline scipy/special/_lambertw.h:52]: In C++14, you can use `1i` to get the imaginary number, which avoids having to call the `std::complex<double>` constructor explicitly. See [complex](https://en.cppreference.com/w/cpp/numeric/complex). I think having C++14 as the minimum standard is totally reasonable here, maybe even C++17. I think PyTorch is on at least C++14.

File: scipy/special/_lambertw.h [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+    } else if (z == 0.0) {
+	if (k == 0) {
+	    return z;
+	}
+	sf_error("lambertw", SF_ERROR_SINGULAR, NULL);
```
Comment (izaid) [inline scipy/special/_lambertw.h:57]: Just noting the `sf_error` call which we'll need to do something about.

File: scipy/special/_lambertw.h [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+     */
+    std::complex<double> w;
+
+    w = std::log(z)
+	+ std::complex<double>(0, 2.0*M_PI*static_cast<double>(k));
+    return w - std::log(w);
+}
+
+
+inline npy_cdouble lambertw_scalar(npy_cdouble zp, long k, double tol) {
```
Comment (izaid) [inline scipy/special/_lambertw.h:161]: You can use `reinterpret_cast` to avoid a lot of boilerplate here. For instance `*reinterpret_cast<std::complex<double> *>(&zp)` will give you a `std::complex<double>`. And you can go the other way as well if you put `npy_cdouble *` in the angular brackets.

**Comment by izaid** [comment]
Comments made! Really happy to see this, great work!

General point is we'll need to think about error handling and do something with `sf_error`. Also, turning these into ufuncs is really not a big deal, can probably get away with a single extra file for all special functions that bridges between C++ and NumPy.

**Comment by izaid** [comment]
@ilayn is right on the money with his unions comment. In C++ language, what you are looking for is `reinterpret_cast`, see my comment above. Yes, it is fine to use that here.

File: scipy/special/_lambertw.h (hunk 4 above)
Comment (izaid) [reply scipy/special/_lambertw.h:31]: Happy to look into this properly with you, but I guess we should probably do it in PR #2.

**Comment by steppi** [comment]
Thanks @izaid, great suggestions. I'll make all the suggested changes tomorrow morning.

File: scipy/special/_lambertw.h (hunk 6 above)
Comment (steppi) [reply scipy/special/_lambertw.h:52]: [Toolchain roadmap](https://docs.scipy.org/doc/scipy/dev/toolchain.html) says C++14 had full support as of 2021, we're not on full C++17 support yet. The constraint is the oldest supported GCC version.

File: scipy/special/_lambertw.h (hunk 4 above)
Comment (steppi) [reply scipy/special/_lambertw.h:31]: yeah, agreed on punting this for later.

File: scipy/special/_lambertw.h (hunk 8 above)
Comment (steppi) [reply scipy/special/_lambertw.h:161]: It works well, but I got this warning

```
../scipy/special/_lambertw.h:161:14: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
  161 |     return *(reinterpret_cast<npy_cdouble*>(&w));
      |             ~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
it should be fine though right. The compiler trusts us not to do anything that breaks strict aliasing, and I think this seems like it does, but doesn't? Is that right?

**Comment by lucascolley** [comment]
@steppi RE: strict aliasing

See the end of http://dbp-consulting.com/tutorials/StrictAliasing.html

tldr; the warning is there because it isn't guaranteed to be portable

I don't know whether this is a problem or whether we can ignore it, but if it is a problem, is `std::memcpy` an alternative?

From the cpp reference Ilhan linked:

_"When it is needed to interpret the bytes of an object as a value of a different type, `std::memcpy` or `std::bit_cast` (since C++20) can be used"_

It looks like we'd like to be able to use `std::bit_cast`, but maybe `std::memcpy` will do instead?

From the blog I linked:

_"You can also accomplish the same thing by using memcpy ... and it will probably be as efficient. Wait, you ask me, how can that be? There will be at least two calls to memcpy added to the mix! Well gcc and other modern compilers have smart optimizers and will, in many cases, ... elide the calls to memcpy."_

**Comment by steppi** [comment]
I've made the suggested changes. The name `_extra_special.h` for the wrappers to use in the ufunc machinery for SciPy is a joke, so hopefully someone else can think of a better name.

Thanks @lucascolley, it looks like the strict aliasing warning lead to a test failure since we had -Werror set. Playing around with %timeit shows no difference anyway, so I think I'll just go back to what I had before.

**Comment by steppi** [comment]
Oh, I see, it's not just a warning. Based on http://dbp-consulting.com/tutorials/StrictAliasing.html, the code the compiler generates when we use `reinterpret_cast` isn't great, so I should still try `memcpy`.

**Comment by steppi** [comment]
> Oh, I see, it's not just a warning. Based on http://dbp-consulting.com/tutorials/StrictAliasing.html, the code the compiler generates when we use `reinterpret_cast` might not be ideal, so I should still try `memcpy`.

No difference from using `memcpy` vs naive copying. If anyone objects to the slight performance decrease for the new implementation, I can play around with godbolt to see what's going on with the generated assembly. Maybe that can be left for the future?

File: scipy/special/_evalpoly.h [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+     * efficient than Horner's method.
+     */
+    int j;
+    double a = coeffs[0];
+    double b = coeffs[1];
+    double r = 2 * z.real();
+    double s = std::norm(z);
+    double tmp;
+
+    for (j = 2; j < degree + 1; j++) {
```
Comment (izaid) [inline scipy/special/_evalpoly.h:34]: Rather then declare all variables at the top, I think it's best practice to declare them when you need them. And I don't believe there is a difference in speed, it just makes things much more readable.

So here you could do `for (int j = 2; ...` and `double tmp = b`.

File: scipy/special/_extra_special.h [hunk 10]
```diff
@@ -0,0 +1,13 @@
+#pragma once
+
+#include <complex>
+#include <numpy/npy_math.h>
+
+#include "_lambertw.h"
+
+
+inline npy_cdouble lambertw_scalar(npy_cdouble zp, long k, double tol) {
```
Comment (izaid) [inline scipy/special/_extra_special.h:9]: I prefer the reinterpret_cast version, but it's not a big deal and we can also just leave it for later. The reason why I prefer the `reinterpret_cast` version is that it can be implemented in a single line, even if there are multiple arguments that need to be cast. I do think that's the C++ way of doing things.

But also it's not a big deal, and I suspect we'll be modifying these helpers later anyway.

File: scipy/special/_lambertw.h [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+}
+
+
+inline std::complex<double> lambertw_asy(std::complex<double> z, long k) {
+    /* Compute the W function using the first two terms of the
+     * asymptotic series. See 4.20 in [1].
+     */
+    std::complex<double> w;
+
+    w = std::log(z) + 2.0*M_PI*static_cast<double>(k)*1i;
```
Comment (izaid) [inline scipy/special/_lambertw.h:73]: Do we need the `static_cast<double>(k)` here? Can we not use just `k`? We might, the complex version of `*` might complain without it, just checking.

File: scipy/special/_lambertw.h [hunk 12]
```diff
@@ -0,0 +1,154 @@
...
+    return w - std::log(w);
+}
+
+
+inline std::complex<double> lambertw(std::complex<double> z, long k, double tol) {
+    int i;
```
Comment (izaid) [inline scipy/special/_lambertw.h:79]: Same point here about not needing to declare these variables at the top, which should declare them in the same lines as their first use.

File: scipy/special/_lambertw.h [hunk 13]
```diff
@@ -0,0 +1,154 @@
...
+    double absz;
+    std::complex<double> w;
+    std::complex<double> ew, wew, wewz, wn;
+
+    if (std::isnan(z.real()) || std::isnan(z.imag())) {
```
Comment (izaid) [inline scipy/special/_lambertw.h:84]: Technically this actually doesn't need to be `else if` at all, it can all be `if` as things get returned. But not a big deal, depends on style.

File: scipy/special/_lambertw.h [hunk 14]
```diff
@@ -0,0 +1,154 @@
...
+		return wn;
+	    } else {
+		w = wn;
+	    }
+	}
+    }
+
+    sf_error("lambertw", SF_ERROR_SLOW,
+	     "iteration failed to converge: %g + %gj", z.real(), z.imag());
+    return std::complex<double>(NAN, NAN);
```
Comment (izaid) [inline scipy/special/_lambertw.h:153]: For `NAN`, I think we should also use `std::numeric_limits<double>`. I think we want `quiet_NaN`.

**Comment by izaid** [comment]
On the type punning stuff, yes, I think `bit_cast` is probably the way to do it in C++20. We're stuck at C++14, so can't use the `std` version. But we can just create our own in a config.h or something, which will probably be needed anyway.

Maybe just leave it as is for now? I'm sure we'll need to work out best practice for what happens in the `extra_special.h` file. Like the name!

**Comment by izaid** [comment]
I do think that the special function code should all be in a namespace to prevent name clashing. Just `namespace scipy { namespace special { ... } }` would be enough. I don't think the functions in `extra_special.h` need to be in a namespace.

File: scipy/special/_evalpoly.h (hunk 9 above)
Comment (ilayn) [reply scipy/special/_evalpoly.h:34]: I'd argue that practice makes things really cumbersome. Having a single place to look is quite comforting when we are dealing with rather long code instead of hunting down where it was defined. 

I think this is one of the strengths of Fortran syntax which is very few. Also Cython has the same structure so we are quite familiar with it.

File: scipy/special/_lambertw.h (hunk 12 above)
Comment (ilayn) [reply scipy/special/_lambertw.h:79]: See above note

File: scipy/special/_evalpoly.h (hunk 9 above)
Comment (steppi) [reply scipy/special/_evalpoly.h:34]: I'm with @ilayn here. This is what we're more used to; I like having all of the variables declared at the top so you can see at a glance what variables will be used with their types. I understand that the standard practice in modern C++ is to declare variables right before they're used. I don't want to be like those people you hear about who "write Fortran in every language" though. Is this a really egregious violation of standard C++ style @izaid, or is it more just a matter of opinion?

File: scipy/special/_extra_special.h (hunk 10 above)
Comment (steppi) [reply scipy/special/_extra_special.h:9]: Yeah, I like `reinterpret_cast` better too. As you saw below, the issue is that in some tests we build with `-Werror -Wstrictaliasing`, so using `reintrepret_cast` breaks the build.

File: scipy/special/_lambertw.h (hunk 11 above)
Comment (steppi) [reply scipy/special/_lambertw.h:73]: Thanks. I checked and it's not actually needed here. You can't directly multiply a `long` and a `std::complex<double>` but since the multiplication is evaluated left to right, that doesn't happen here. In an earlier version of the translation I was directly multiplying `k` with a complex value, got an error, and just blindly threw the `static_cast`s in everywhere to appease the compiler.

File: scipy/special/_lambertw.h (hunk 13 above)
Comment (steppi) [reply scipy/special/_lambertw.h:84]: Right. I usually try to avoid `else` when the previous branch in the conditional returns. This is just a consequence of mechanically translating the original Cython line by line.

**Comment by steppi** [comment]
> I do think that the special function code should all be in a namespace to prevent name clashing. Just `namespace scipy { namespace special { ... } }` would be enough. I don't think the functions in `extra_special.h` need to be in a namespace.

Good idea. Done. Should I put the helper functions in an anonymous namespace to make them private, or is that unnecessary?

**Comment by steppi** [comment]
Not for this PR, but for the problem of using `sf_error.h`, does anyone have any thoughts on having a preprocessor symbol like `USE_SF_ERROR` and then doing stuff like

```C++
#ifdef USE_SF_ERROR
#include <sf_error.h>
#endif
```

```C++
#ifdef USE_SF_ERROR
sf_error("lambertw", SF_ERROR_SINGULAR, NULL);
#endif
```

File: scipy/special/_evalpoly.h (hunk 9 above)
Comment (izaid) [reply scipy/special/_evalpoly.h:34]: I wouldn't say it's egregious. But... :)

I would say what I described is the much more common style. Indeed, that's what's written in [ISOCpp](https://isocpp.org/wiki/faq/coding-standards) and [Google's C++ style guide](https://google.github.io/styleguide/cppguide.html). Readable is only one concern, there is an efficiency one as well. Empty declarations like those will also call the default constructor, though not for primitive types. Also certain types don't have a default constructor, and therefore can be used only when they are initialised. This includes references, e.g. `double &x` won't compile. Those are the reasons why it's more common to do it that way...

Anyway, that's my two cents!

File: scipy/special/_extra_special.h (hunk 10 above)
Comment (izaid) [reply scipy/special/_extra_special.h:9]: Yep, I saw! What we can think about the best way to do this, is we'll have to do it again and again for all functions we wrap. What I'd suggest is we come up with our own `bit_cast` until C++20 is workable (a long time), but can do that in a later PR.

File: scipy/special/_lambertw.h (hunk 11 above)
Comment (izaid) [reply scipy/special/_lambertw.h]: Nice! Yes, I remember `std::complex<double>` plays badly with integers sometimes.

File: scipy/special/_lambertw.h (hunk 13 above)
Comment (izaid) [reply scipy/special/_lambertw.h]: That's what I thought regarding the translation. Don't feel strongly, whatever you think is best!

**Comment by izaid** [comment]
> > I do think that the special function code should all be in a namespace to prevent name clashing. Just `namespace scipy { namespace special { ... } }` would be enough. I don't think the functions in `extra_special.h` need to be in a namespace.
> 
> Good idea. Done. Should I put the helper functions in an anonymous namespace to make them private, or is that unnecessary?

Do you mean the ones in `extra_special.h`? If yes, I wouldn't worry about them, as I think they'll only be used by SciPy.

**Comment by izaid** [comment]
> Not for this PR, but for the problem of using `sf_error.h`, does anyone have any thoughts on having a preprocessor symbol like `USE_SF_ERROR` and then doing stuff like
> 
> ```c++
> #ifdef USE_SF_ERROR
> #include <sf_error.h>
> #endif
> ```
> 
> ```c++
> #ifdef USE_SF_ERROR
> sf_error("lambertw", SF_ERROR_SINGULAR, NULL);
> #endif
> ```

Yes, something like this would work. What I think is a bit better is to redefine `sf_error` like:

```
ReturnType sf_error(ArgTypes... args) {
#ifdef USE_SF_ERROR
// do stuff
#else
// do nothing
#endif 
}
```

This way we can leave the actual function calls in place, but nothing will happen on systems that don't support it.

**Comment by steppi** [comment]
> > > I do think that the special function code should all be in a namespace to prevent name clashing. Just `namespace scipy { namespace special { ... } }` would be enough. I don't think the functions in `extra_special.h` need to be in a namespace.
> > 
> > 
> > Good idea. Done. Should I put the helper functions in an anonymous namespace to make them private, or is that unnecessary?
> 
> Do you mean the ones in `extra_special.h`? If yes, I wouldn't worry about them, as I think they'll only be used by SciPy.

Sorry, I meant stuff like `lambertw_branchpt` and `lambertw_pade0`. The things used internally by the main `lambertw` implementatoin.

File: scipy/special/_evalpoly.h (hunk 9 above)
Comment (steppi) [reply scipy/special/_evalpoly.h]: I changed it to declare the loop indices at first use, and fixed a place where I had an empty declaration and then put the definition on the next line. The variable `absz` could  be changed to be declared at first use in `lambertw` in `_lambertw.h`, but I don't think any of the others could be changed without torturing the code in some way, since each of `w`, `wz`, `wew` and `wewz` and `wn` are defined deep in nested logic, and in different ways in different branches.

Comment (steppi) [reply scipy/special/_evalpoly.h]: Again for `cevalpoly`, setting `double tmp = b` before the loop, would require moving `tmp = b` to the end of the loop, and if we didn't want to set `tmp` one more time than needed, we'd need to stop the loop one early, and then put 

`C++
	b = std::fma(-s, a, coeffs[j]);
	a = std::fma(r, a, tmp);
```

after the loop, which just doesn't seem worth it. I don't ike setting `tmp = b` an extra time either, even though it probably has no real performance impact.

**Comment by izaid** [comment]
> > > > I do think that the special function code should all be in a namespace to prevent name clashing. Just `namespace scipy { namespace special { ... } }` would be enough. I don't think the functions in `extra_special.h` need to be in a namespace.
> > > 
> > > 
> > > Good idea. Done. Should I put the helper functions in an anonymous namespace to make them private, or is that unnecessary?
> > 
> > 
> > Do you mean the ones in `extra_special.h`? If yes, I wouldn't worry about them, as I think they'll only be used by SciPy.
> 
> Sorry, I meant stuff like `lambertw_branchpt` and `lambertw_pade0`. The things used internally by the main `lambertw` implementatoin.

I don't think this matters so much right now, but it might be something we have to do at some point. Another common option is a `detail` namespace, i.e. `namespace scipy { namespace special { namespace detail { ... } } }`. But I'd probably just leave it alone for now.

File: scipy/special/_evalpoly.h (hunk 9 above)
Comment (ilayn) [reply scipy/special/_evalpoly.h]: Well we are 1 function in and we already think the standard is rubbish so I guess nothing has changes since I stopped using C/C++ 😄 

Joking aside, I don't think this can actually matter since these are just a few bytes and not like large shaders or whatever. I'm pretty confident the standard is not talking to us with that paragraph.

**Review by izaid — APPROVED** [review]

**Comment by izaid** [comment]
Happy with this, have approved! Excellent work. :)

Going forward, we'll need to tackle the `sf_error` header before we try anything in the other array libraries. Think that should be next up.

**Comment by steppi** [comment]
@tirthasheshpatel do you want to make a final pass and merge if you think it's ready?

File: scipy/special/_lambertw.h [hunk 15]
```diff
@@ -0,0 +1,162 @@
...
+ * Distributed under the same license as SciPy
+ * Translated into C++ by SciPy developers, 2023.
+ *
+ * References:
+ * [1] On the Lambert W function, Adv. Comp. Math. 5 (1996) 329-359,
+ *     available online: http://www.apmaths.uwo.ca/~djeffrey/Offprints/W-adv-cm.pdf
+ * [2] mpmath source code, Subversion revision 990
+ *     http://code.google.com/p/mpmath/source/browse/trunk/mpmath/functions.py?spec=svn994&r=992
+ * [3] mpmath source code, Subversion revision 994
+ *     http://code.google.com/p/mpmath/source/browse/trunk/mpmath/function_docs.py?spec=svn994&r=994
```
Comment (j-bowhay) [inline scipy/special/_lambertw.h:16]: Not a blocking comment but these references no longer work

Comment (steppi) [reply scipy/special/_lambertw.h:16]: Thanks! Good catch

Comment (steppi) [reply scipy/special/_lambertw.h:16]: Maybe just link to the current commits on github for these files? i looked through the history and it doesn't seem like these implementations have been touched in the past decade or more.

https://github.com/mpmath/mpmath/blob/c5939823669e1bcce151d89261b802fe0d8978b4/mpmath/functions/functions.py

and

https://github.com/mpmath/mpmath/blob/c5939823669e1bcce151d89261b802fe0d8978b4/mpmath/function_docs.py

Comment (steppi) [reply scipy/special/_lambertw.h]: I updated the links, using internet archive where possible.

**Review by tirthasheshpatel — APPROVED** [review]
I confirmed that the C++ implementation matches the Cython implementation.

So, LGTM! Excited to see how this can be run on CUDA and/or with other array frameworks!

Thanks @steppi @izaid for the amazing work!
